server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -3568,58 +3569,96 @@ private EngineConfig newEngineConfig(LongSupplier globalCheckpointSupplier) {
         );
     }
 
+    /**
+     * Check to run before running the primary permit operation
+     */
+    public enum PrimaryPermitCheck {
```

Comment:
Yes this is an approach. A bit over-engineering just to avoid an assertion :D but ok! I am wondering if we're opening a can of worms though in the sense that other people might be tempted to use the NONE PrimaryPermitCheck wrongly. Maybe lets enrich the comment to say that for the moment NONE is used during the recovery of hollow shards and warn people against using NONE if they're unsure.

Reply:
Added a comment warning about not disabling primary permit checks without a reason.